File: app/src/main/java/com/example/myapplication/core/nutrition/NutritionXlsxParser.kt
package com.example.myapplication.core.nutrition

import org.xmlpull.v1.XmlPullParser
import org.xmlpull.v1.XmlPullParserFactory
import java.io.ByteArrayInputStream
import java.io.InputStream
import java.util.zip.ZipInputStream

object NutritionXlsxParser {
    fun parse(xlsxBytes: ByteArray, batchId: String = ""): CsvParseResult {
        return try {
            val rows = extractRowsFromXlsx(xlsxBytes)
            NutritionCsvParser.parseTable(rows, batchId)
        } catch (e: Exception) {
            CsvParseResult(emptyList(), listOf("Lỗi đọc tệp Excel: ${e.message}"))
        }
    }

    private fun extractRowsFromXlsx(xlsxBytes: ByteArray): List<List<String>> {
        var sharedStrings = listOf<String>()
        var sheetRows = listOf<List<String>>()

        ZipInputStream(ByteArrayInputStream(xlsxBytes)).use { zip ->
            var entry = zip.nextEntry
            while (entry != null) {
                if (entry.name == "xl/sharedStrings.xml") {
                    val entryBytes = zip.readBytes()
                    sharedStrings = parseSharedStrings(ByteArrayInputStream(entryBytes))
                } else if (entry.name == "xl/worksheets/sheet1.xml") {
                    val entryBytes = zip.readBytes()
                    sheetRows = parseSheetRows(ByteArrayInputStream(entryBytes), sharedStrings)
                }
                entry = zip.nextEntry
            }
        }
        return sheetRows
    }

    private fun parseSharedStrings(inputStream: InputStream): List<String> {
        val strings = mutableListOf<String>()
        val factory = XmlPullParserFactory.newInstance()
        val parser = factory.newPullParser()
        parser.setInput(inputStream, "UTF-8")

        var eventType = parser.eventType
        var inT = false
        val sb = java.lang.StringBuilder()

        while (eventType != XmlPullParser.END_DOCUMENT) {
            when (eventType) {
                XmlPullParser.START_TAG -> {
                    if (parser.name == "t") {
                        inT = true
                        sb.setLength(0)
                    }
                }
                XmlPullParser.TEXT -> {
                    if (inT) {
                        sb.append(parser.text)
                    }
                }
                XmlPullParser.END_TAG -> {
                    if (parser.name == "t") {
                        inT = false
                    } else if (parser.name == "si") {
                        strings.add(decodeXmlEntities(sb.toString()))
                    }
                }
            }
            eventType = parser.next()
        }
        return strings
    }

    private fun parseSheetRows(inputStream: InputStream, sharedStrings: List<String>): List<List<String>> {
        val rows = mutableListOf<List<String>>()
        val factory = XmlPullParserFactory.newInstance()
        val parser = factory.newPullParser()
        parser.setInput(inputStream, "UTF-8")

        var eventType = parser.eventType
        var currentColumnIndex = -1
        var isShared = false
        var cellValue: String? = null
        val currentCells = mutableMapOf<Int, String>()
        var maxColIndex = -1

        while (eventType != XmlPullParser.END_DOCUMENT) {
            when (eventType) {
                XmlPullParser.START_TAG -> {
                    val tagName = parser.name
                    if (tagName == "row") {
                        currentCells.clear()
                        maxColIndex = -1
                    } else if (tagName == "c") {
                        val rAttr = parser.getAttributeValue(null, "r")
                        currentColumnIndex = if (rAttr != null) getColumnIndex(rAttr) else -1
                        if (currentColumnIndex > maxColIndex) {
                            maxColIndex = currentColumnIndex
                        }
                        val tAttr = parser.getAttributeValue(null, "t")
                        isShared = tAttr == "s"
                        cellValue = null
                    } else if (tagName == "v" || tagName == "t") {
                        cellValue = ""
                    }
                }
                XmlPullParser.TEXT -> {
                    if (cellValue != null) {
                        cellValue += parser.text
                    }
                }
                XmlPullParser.END_TAG -> {
                    val tagName = parser.name
                    if (tagName == "c") {
                        if (currentColumnIndex != -1) {
                            val finalVal = if (isShared && cellValue != null) {
                                val idx = cellValue!!.toIntOrNull()
                                if (idx != null && idx >= 0 && idx < sharedStrings.size) {
                                    sharedStrings[idx]
                                } else {
                                    ""
                                }
                            } else {
                                cellValue ?: ""
                            }
                            currentCells[currentColumnIndex] = decodeXmlEntities(finalVal)
                        }
                    } else if (tagName == "row") {
                        val result = MutableList(maxColIndex + 1) { "" }
                        for ((col, valStr) in currentCells) {
                            result[col] = valStr
                        }
                        rows.add(result)
                    }
                }
            }
            eventType = parser.next()
        }
        return rows
    }

    private fun getColumnIndex(ref: String): Int {
        val letters = ref.takeWhile { it.isLetter() }.uppercase()
        var col = 0
        for (char in letters) {
            col = col * 26 + (char - 'A' + 1)
        }
        return col - 1
    }

    private fun decodeXmlEntities(str: String): String {
        if (!str.contains('&')) return str
        return str
            .replace("&amp;", "&")
            .replace("&lt;", "<")
            .replace("&gt;", ">")
            .replace("&quot;", "\"")
            .replace("&apos;", "'")
    }
}

File: flutter/lib/core/nutrition/nutrition_xlsx_parser.dart
import 'dart:convert';
import 'dart:typed_data';
import 'package:archive/archive.dart';
import 'package:xml/xml.dart';
import 'nutrition_csv_parser.dart';

class NutritionXlsxParser {
  static CsvParseResult parse(Uint8List xlsxBytes, {String batchId = ""}) {
    try {
      final rows = _extractRowsFromXlsx(xlsxBytes);
      return NutritionCsvParser.parseTable(rows, batchId: batchId);
    } catch (e) {
      return CsvParseResult(items: [], warnings: ["Lỗi đọc tệp Excel: $e"]);
    }
  }

  static List<List<String>> _extractRowsFromXlsx(Uint8List xlsxBytes) {
    final archive = ZipDecoder().decodeBytes(xlsxBytes);
    List<String> sharedStrings = [];
    List<List<String>> sheetRows = [];

    ArchiveFile? sharedStringsFile;
    ArchiveFile? sheetFile;

    for (final file in archive) {
      if (file.name == 'xl/sharedStrings.xml') {
        sharedStringsFile = file;
      } else if (file.name == 'xl/worksheets/sheet1.xml') {
        sheetFile = file;
      }
    }

    if (sharedStringsFile != null) {
      final content = sharedStringsFile.content as List<int>;
      final xmlContent = utf8.decode(content);
      sharedStrings = _parseSharedStrings(xmlContent);
    }

    if (sheetFile != null) {
      final content = sheetFile.content as List<int>;
      final xmlContent = utf8.decode(content);
      sheetRows = _parseSheetRows(xmlContent, sharedStrings);
    }

    return sheetRows;
  }

  static List<String> _parseSharedStrings(String xmlContent) {
    final document = XmlDocument.parse(xmlContent);
    final strings = <String>[];
    final siElements = document.findAllElements('si');
    for (final si in siElements) {
      final tElements = si.findAllElements('t');
      final sb = StringBuffer();
      for (final t in tElements) {
        sb.write(t.innerText);
      }
      strings.add(_decodeXmlEntities(sb.toString()));
    }
    return strings;
  }

  static List<List<String>> _parseSheetRows(String xmlContent, List<String> sharedStrings) {
    final document = XmlDocument.parse(xmlContent);
    final rows = <List<String>>[];
    final rowElements = document.findAllElements('row');

    for (final row in rowElements) {
      final cellMap = <int, String>{};
      var maxColIndex = -1;

      final cellElements = row.findAllElements('c');
      for (final cell in cellElements) {
        final rAttr = cell.getAttribute('r');
        final currentColumnIndex = rAttr != null ? _getColumnIndex(rAttr) : -1;
        if (currentColumnIndex > maxColIndex) {
          maxColIndex = currentColumnIndex;
        }

        final tAttr = cell.getAttribute('t');
        final isShared = tAttr == 's';

        final vElement = cell.findElements('v').firstOrNull;
        var cellValue = vElement?.innerText ?? "";

        if (isShared && cellValue.isNotEmpty) {
          final idx = int.tryParse(cellValue);
          if (idx != null && idx >= 0 && idx < sharedStrings.length) {
            cellValue = sharedStrings[idx];
          } else {
            cellValue = "";
          }
        }

        if (currentColumnIndex != -1) {
          cellMap[currentColumnIndex] = _decodeXmlEntities(cellValue);
        }
      }

      final rowCells = List.filled(maxColIndex + 1, "");
      cellMap.forEach((col, val) {
        rowCells[col] = val;
      });
      rows.add(rowCells);
    }
    return rows;
  }

  static int _getColumnIndex(String ref) {
    final letters = ref.split(RegExp(r'[0-9]')).first.toUpperCase();
    var col = 0;
    for (var i = 0; i < letters.length; i++) {
      final charCode = letters.codeUnitAt(i);
      col = col * 26 + (charCode - 65 + 1);
    }
    return col - 1;
  }

  static String _decodeXmlEntities(String str) {
    if (!str.contains('&')) return str;
    return str
        .replaceAll('&amp;', '&')
        .replaceAll('&lt;', '<')
        .replaceAll('&gt;', '>')
        .replaceAll('&quot;', '"')
        .replaceAll('&apos;', "'");
  }
}
